File: src/012/p012.exs
#!/usr/bin/env elixir
defmodule Problem012 do
  defp numbers(), do: Stream.unfold(1, fn(x) -> {x, x+1} end)

  defp triangle_number(x), do: (1..x) |> Enum.sum

  defp factors(n) do
    1..round(:math.sqrt(n))
    |> Enum.filter(fn(x) -> rem(n, x) == 0 end)
  end

  defp condition(x) do
    x
    |> triangle_number
    |> factors
    |> Enum.count
    |> Kernel.*(2) < 500
  end

  def solve do
    numbers()
    |> Stream.take_while(&(condition(&1)))
    |> Enum.max
    |> Kernel.+(1)
    |> triangle_number
  end
end

IO.puts Problem012.solve

File: src/012/p012.rb
class Integer
  def triangle_number
    (1..self).reduce(:+)
  end

  def factors
    (1..Math.sqrt(self)).select { |i| (self % i).zero? }.each_with_object([]) do |number, list|
      list << number
      list << self / number unless self / number == number
    end
  end
end

def solve
  x = 1
  x += 1 while x.triangle_number.factors.count < 500
  x.triangle_number
end

puts solve
